### ISSUE TITLE
Add optional `allow_single_version` Boolean arg to `actx.compile`

### ISSUE THREAD

**Issue by inducer** [inducer/arraycontext#227]
Benefits:

- Avoids unexpected repeated compilation
- Avoid dispatch cost to select which implementation to use

Drawbacks:

- Less flexible, of course :)

Naming of the flag is an open question.

cc @majosm @mtcam

**Comment by kaushikcfd** [comment]
I don't think I understand this. Could you please elaborate? Does this have to do with sub-expressions being semantically same but have different metadata?
Thanks!

**Comment by inducer** [comment]
Sorry, I didn't spend enough time explaining. Right now, we have reasonably elaborate dispatch logic between different versions of a compiled function, based on shapes, which arguments are supplied, etc. This flag would assert that only a single version should ever exist, and calling the compiled function a second time with mismatched arguments would be an error. Does that make sense?

**Comment by kaushikcfd** [comment]
Yep, makes sense. Thanks! Essentially this flag should ensure that length of [`BaseLazilyCompilingFunctionCaller.program`](https://github.com/inducer/arraycontext/blob/829f8929a82afd874f4f598341866e3ab0ec1dcd/arraycontext/impl/pytato/compile.py#L262-L263) should not exceed 1.

**Comment by inducer** [comment]
Yup, exactly.

**Comment by majosm** [comment]
Question: Would the `allow_single_version=True` case check if the arguments are the same as when it was compiled? Or is it more of an "at your own risk" thing?

**Comment by kaushikcfd** [comment]
> Question: Would the allow_single_version=True case check if the arguments are the same as when it was compiled? Or is it more of an "at your own risk" thing?

It should be an error at [this point](https://github.com/inducer/arraycontext/blob/829f8929a82afd874f4f598341866e3ab0ec1dcd/arraycontext/impl/pytato/compile.py#L327) if `len(self.program_cache) > 1`.

**Comment by majosm** [comment]
> > Question: Would the allow_single_version=True case check if the arguments are the same as when it was compiled? Or is it more of an "at your own risk" thing?
> 
> It should be an error at [this point](https://github.com/inducer/arraycontext/blob/829f8929a82afd874f4f598341866e3ab0ec1dcd/arraycontext/impl/pytato/compile.py#L327) if `len(self.program_cache) > 1`.

Does it still incur the "dispatch cost to select which implementation to use" then?

**Comment by kaushikcfd** [comment]
Aah fair. I missed that as a constraint. Though I am unsure how to avoid the dispatch cost since we would always need to compute the `arg_id_to_descr`.

**Comment by majosm** [comment]
In any case, I can see the value in avoiding multiple compiles. On the user side, though, I'm not sure if it's always obvious enough when it's happening that one would know to intervene. Kind of makes me wonder if the `single_version_only=True` behavior should be the default? i.e., emit an error if it tries to compile multiple times, unless the caller explicitly says it's OK. Not sure how much stuff this would break, though.

**Comment by inducer** [comment]
At the very least, it would break the LSRK45 time stepper, which relies on having two different versions at the moment.

https://github.com/inducer/grudge/blob/c9906c9a20bfb5c78733093ccba716d7ac036897/grudge/shortcuts.py#L44-L62

**Comment by matthiasdiener** [comment]
How about just emitting a warning instead of an exception? This would even work without necessarily changing the API.

**Comment by inducer** [comment]
IMO, both multi-version and single-version are legitimate uses of the function, unless the user says otherwise.
